lib/strings.go

```diff
@@ -1043,3 +1045,70 @@ func (l stringLib) validString(arg ref.Val) ref.Val {
 	}
 	return types.DefaultTypeAdapter.NativeToValue(utf8.Valid([]byte(s)))
 }
+
+// Printf returns a cel.EnvOption to configure an extended function for
+// formatting strings and associated arguments.
+//
+// # Sprintf
+//
+// Formats a string from a format string and a set of arguments using the
+// Go fmt syntax:
+//
+//	<string>.sprintf(<list<dyn>>) -> <string>
+//	sprintf(<string>, <list<dyn>>) -> <string>
+//
+// Since CEL does not support variadic functions, arguments are provided
+// as an array corresponding to the normal Go variadic slice.
+//
+// Examples:
+//
+//	"Hello, %s".sprintf(["World!"])   // return "Hello, World!"
+//	sprintf("Hello, %s", ["World!"])  // return "Hello, World!"
+func Printf() cel.EnvOption {
+	return cel.Lib(printfLib{})
+}
+
+type printfLib struct{}
+
+func (l printfLib) CompileOptions() []cel.EnvOption {
+	return []cel.EnvOption{
+		cel.Function("sprintf",
+			cel.MemberOverload(
+				"sprintf_string_list_string",
+				[]*cel.Type{types.StringType, listDyn},
+				types.StringType,
+				cel.BinaryBinding(l.sprintf),
+			),
+			cel.Overload(
+				"string_sprintf_list_string",
+				[]*cel.Type{types.StringType, listDyn},
+				types.StringType,
+				cel.BinaryBinding(l.sprintf),
+			),
+		),
+	}
+}
+
+func (l printfLib) ProgramOptions() []cel.ProgramOption { return nil }
+
+func (l printfLib) sprintf(arg0, arg1 ref.Val) ref.Val {
+	format, ok := arg0.(types.String)
+	if !ok {
+		return types.ValOrErr(format, "no such overload for sprintf: format is not a string")
+	}
+	argList, ok := arg1.(traits.Lister)
+	if !ok {
+		return types.ValOrErr(argList, "no such overload for sprintf: args is not a list")
+	}
+	n, _ := argList.Size().(types.Int) // Continue since this is an optimisation.
+	args := make([]any, 0, n)
+	it := argList.Iterator()
+	for i := 1; it.HasNext() == types.True; i++ {
+		v, err := it.Next().ConvertToNative(reflectAnyType)
+		if err != nil {
+			return types.NewErr("failed to convert argument %d: %v", i, err)
+		}
+		args = append(args, v)
+	}
+	return types.String(fmt.Sprintf(string(format), args...))
```

Comment:
Since in L#1095: `format, ok := arg0.(types.String)` we already asserted the `format` is a string, is it required to convert it again in L#1113?

Reply:
We assert it to a `types.String` which is a named type with a underlying type `string`. We cannot use a `type.String` in the place of a `string` since although they have the same underlying type, the spec only allows assignment between different types without conversion if one of the types is not named. This is detailed in the [Assignability](https://go.dev/ref/spec#Assignability) section of the spec.

> V and T have identical [underlying types](https://go.dev/ref/spec#Underlying_types) but are not type parameters and at least one of V or T is not a [named type](https://go.dev/ref/spec#Types).

The situation in the use of the `types.Int` in the `make` call is less constrained because `make`'s integer arguments are not specified to be `int`, but rather [`builtin.IntegerType`](https://pkg.go.dev/builtin#make).